paperscraper/pdf.py

```diff
@@ -157,8 +191,277 @@ def save_pdf_from_dump(
             logger.warning(f"Skipping {paper['title']} since no DOI available.")
             continue
         filename = paper[key_to_save].replace("/", "_")
+        pdf_file = Path(os.path.join(pdf_path, f"{filename}.pdf"))
+        xml_file = pdf_file.with_suffix(".xml")
+        if pdf_file.exists():
+            logger.info(f"File {pdf_file} already exists. Skipping download.")
+            continue
+        if xml_file.exists():
+            logger.info(f"File {xml_file} already exists. Skipping download.")
+            continue
+        output_path = str(pdf_file)
         save_pdf(
             paper,
-            os.path.join(pdf_path, f"{filename}.pdf"),
+            output_path,
             save_metadata=save_metadata,
+            api_keys=api_keys_dict
         )
+
+def load_api_keys(filepath):
+    """
+    Reads API keys from a file and returns them as a dictionary.
+    The file should have each API key on a separate line in the format:
+        KEY_NAME=API_KEY_VALUE
+
+    Example:
+        WILEY_TDM_API_TOKEN=your_wiley_token_here
+        ELSEVIER_TDM_API_KEY=your_elsevier_key_here
+
+    Args:
+        filepath (str): Path to the file containing API keys.
+
+    Returns:
+        Dict[str, str]: A dictionary where keys are API key names and values are their respective API keys.
+    """
+    api_keys = {}
+    try:
+        with open(filepath, "r") as f:
+            for line in f:
+                if "=" in line:
+                    key, value = line.strip().split("=", 1)
+                    api_keys[key] = value
+    except Exception as e:
+        print(f"Error reading API keys file: {e}")
+    return api_keys
+
+def fallback_wiley_api(paper_metadata: Dict[str, Any], output_path: Path, logger: logging.Logger, api_keys: Dict[str, str]):
+    """
+    Attempt to download the PDF via the Wiley TDM API (popular publisher which blocks standard scraping attempts; API access free for academic users).
+
+    This function uses the WILEY_TDM_API_TOKEN environment variable to authenticate
+    with the Wiley TDM API and attempts to download the PDF for the given paper.
+    See https://onlinelibrary.wiley.com/library-info/resources/text-and-datamining for a description on how to get your WILEY_TDM_API_TOKEN.
+
+    Args:
+        paper_metadata (dict): Dictionary containing paper metadata. Must include the 'doi' key.
+        output_path (Path): A pathlib.Path object representing the path where the PDF will be saved.
+        logger (logging.Logger): Logger instance used for logging information and errors.
+        api_keys (dict): Preloaded API keys.
+    """
+
+    WILEY_TDM_API_TOKEN = api_keys.get("WILEY_TDM_API_TOKEN")
+    encoded_doi = paper_metadata["doi"].replace("/", "%2F")
+    api_url = f"https://api.wiley.com/onlinelibrary/tdm/v1/articles/{encoded_doi}"
+    headers = {"Wiley-TDM-Client-Token": WILEY_TDM_API_TOKEN}
+
+    max_attempts = 2
```

Comment:
This should be made controllable as argument. Default can be 2

Reply:
I added it as argument to the wiley API function. 
Passing it down from higher up (e.g. save_pdf) is probably less useful, as the max_attempts is specific to the wiley API (it happens that it fails and then works second time), while I haven't seen other APIs fail without reason (hence they have no max_attempts argument, since no repeat function is implemented).